### PR TITLE
expression: let `cast` function supports explicit set charset

### LINKED ISSUE #55677: Pushdown TopN into TiKV with `cast` function meets incorrect result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE `test` (
  `id` bigint(20) NOT NULL,
  `update_user` varchar(32) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into test values(1,'张三');
insert into test values(2,'李四');
insert into test values(3,'张三');
insert into test values(4,'李四');
insert into test values(5,'张三');
insert into test values(6,'李四');
insert into test values(7,'张三');
insert into test values(8,'李四');
insert into test values(9,'张三');
insert into test values(10,'李四');
insert into test values(11,'张三');
insert into test values(12,'李四');
insert into test values(13,'张三');
insert into test values(14,'李四'); 
select * from test order by cast(update_user as char character set gbk) desc , id limit 10; 
```

Another test case
```
CREATE TABLE `test` (
  `id` bigint NOT NULL,
  `update_user` varchar(32) CHARACTER SET gbk COLLATE gbk_chinese_ci DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into test values(1,'张三');
insert into test values(2,'李四');
insert into test values(3,'张三');
insert into test values(4,'李四');
insert into test values(5,'张三');
insert into test values(6,'李四');
insert into test values(7,'张三');
insert into test values(8,'李四');
insert into test values(9,'张三');
insert into test values(10,'李四');
insert into test values(11,'张三');
insert into test values(12,'李四');
insert into test values(13,'张三');
insert into test values(14,'李四'); 
select * from test order by cast(update_user as char) desc , id limit 10; 
```

### 2. What did you expect to see? (Required)

```
MySQL [test]>  select * from test order by cast(update_user as char character set gbk) desc , id limit 10;   
+----+-------------+
| id | update_user |
+----+-------------+
|  1 | 张三        |
|  3 | 张三        |
|  5 | 张三        |
|  7 | 张三        |
|  9 | 张三        |
| 11 | 张三        |
| 13 | 张三        |
|  2 | 李四        |
|  4 | 李四        |
|  6 | 李四        |
+----+-------------+
10 rows in set (0.00 sec)
```

Second case
```
mysql> select * from test order by cast(update_user as char) desc , id limit 10;
+----+-------------+
| id | update_user |
+----+-------------+
|  2 | 李四        |
|  4 | 李四        |
|  6 | 李四        |
|  8 | 李四        |
| 10 | 李四        |
| 12 | 李四        |
| 14 | 李四        |
|  1 | 张三        |
|  3 | 张三        |
|  5 | 张三        |
+----+-------------+
10 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)

```
MySQL [test]>  select * from test order by cast(update_user as char character set gbk) desc, id limit 10; 
+----+-------------+
| id | update_user |
+----+-------------+
|  1 | 张三        |
|  3 | 张三        |
|  5 | 张三        |
|  2 | 李四        |
|  4 | 李四        |
|  6 | 李四        |
|  8 | 李四        |
| 10 | 李四        |
| 12 | 李四        |
| 14 | 李四        |
+----+-------------+
10 rows in set (0.00 sec)
```

Second case
```
mysql> select * from test order by cast(update_user as char) desc , id limit 10;
+----+-------------+
| id | update_user |
+----+-------------+
|  2 | 李四        |
|  4 | 李四        |
|  6 | 李四        |
|  1 | 张三        |
|  3 | 张三        |
|  5 | 张三        |
|  7 | 张三        |
|  9 | 张三        |
| 11 | 张三        |
| 13 | 张三        |
+----+-------------+
10 rows in set (25.76 sec)
```

### 4. What is your TiDB version? (Required)

v7.5.1, nightly(v8.3.0+) and more

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #55677

Problem Summary: Ref https://dev.mysql.com/doc/refman/8.4/en/cast-functions.html#cast-function-descriptions

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/expression/builtin_cast.go [hunk 1]
```diff
@@ -288,14 +288,15 @@ func (c *castAsDecimalFunctionClass) getFunction(ctx BuildContext, args []Expres
 type castAsStringFunctionClass struct {
 	baseFunctionClass
 
-	tp *types.FieldType
+	tp                *types.FieldType
+	isExplicitCharSet bool
```
Comment (XuHuaiyu) [inline pkg/expression/builtin_cast.go:292]: CharSet -> Charset ?

Comment (Defined2014) [reply pkg/expression/builtin_cast.go]: Done

**Review by tiancaiamao — APPROVED** [review]

File: pkg/expression/collation.go [hunk 2]
```diff
@@ -279,9 +297,8 @@ func deriveCollation(ctx BuildContext, funcName string, args []Expression, retTy
 	case ast.Cast:
 		// We assume all the cast are implicit.
 		ec = &ExprCollation{args[0].Coercibility(), args[0].Repertoire(), args[0].GetType(ctx.GetEvalCtx()).GetCharset(), args[0].GetType(ctx.GetEvalCtx()).GetCollate()}
-		// Non-string type cast to string type should use @@character_set_connection and @@collation_connection.
-		// String type cast to string type should keep its original charset and collation. It should not happen.
-		if retType == types.ETString && argTps[0] != types.ETString {
+		// Cast to string type should use @@character_set_connection and @@collation_connection.
+		if retType == types.ETString {
```
Comment (Defined2014) [inline pkg/expression/collation.go:301]: When charset is implicit, cast to char use default charset and collation

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**